Splitting CheckInvPaste() into smaller functions

matheusgomes28 opened the pull request:
Moved functions that were getting previous items into smaller ones with better names. At least you can see straight-away what each of these blocks of code do!

This has been sitting on my github for a few weeks, just got back to this today and realise the `timedemo` test is failing. Will fix that, but creating this to get some feedback anyway.

matheusgomes28 commented:
I am struggling a little bit to debug the `timedemo_test`, as I get the following errors when running it locally:

```
ERROR: Asset not found: lua\user.lua
Demo file not found
...
```

I can't find `user.lua` anywhere, am I missing something?

StephenCWills commented:
Are you debugging with Visual Studio?

matheusgomes28 commented:
Thanks for the reply, @StephenCWills .

I am running on VsCode with the MSVC compilers on Windows, and the clang compilers on Ubuntu. I figured what the issue I have is: it's simply that we get the `prefPath` in `paths.cpp` from the SDL call to `SDL_GetBasePath()` which get's where the binary is located. Given my following `launch.json` in VScode:

```json
        {
            "name": "Timedemo Debug",
            "type": "cppvsdbg",
            "request": "launch",
            "program": "${workspaceFolder}\\build\\vs2022-deb\\Debug\\timedemo_test.exe",
            "cwd": "${workspaceFolder}\\build\\vs2022-deb\\",
            "args": [
                "--config-dir",
                "C:\\Users\\mathe\\development\\devilutionX\\build\\vs2022-deb\\test\\fixtures\\timedemo\\WarriorLevel1to2",
                "--save-dir",
                "C:\\Users\\mathe\\development\\devilutionX\\build\\vs2022-deb\\test\\fixtures\\timedemo\\WarriorLevel1to2",
                "--diablo",
                "--spawn",
                "--lang",
                "en",
                "--demo",
                "0",
                "--timedemo"
            ],
            "stopAtEntry": false,
            "environment": [],
            "console": "integratedTerminal"
        }
```

`SDL_GetBasePath()` will return `${workspaceFolder}\\build\\vs2022-deb\\Debug` as the base path, and we later add on the paths `text/fixture/...` to that in order to create the path to `demo_0.dmo`. This works fine when you have ctest integration in your IDE (like visual studio, clion, or simply running `ctest`) because [CMake seems to copy the bin as if it was called from the directory where `add_test(...)` was called from](https://cmake.org/cmake/help/latest/prop_test/WORKING_DIRECTORY.html).

I do wonder: should we not use the value of the current working directory as the base path, instead of the path to the binary? Either way, I haven't fixed it yet, and the solution seems to just run the test through the test explorer and let `ctest` do the magic.

StephenCWills commented:
> `SDL_GetBasePath()` will return `${workspaceFolder}\\build\\vs2022-deb\\Debug` as the base path, and we later add on the paths `text/fixture/...` to that in order to create the path to `demo_0.dmo`. This works fine when you have ctest integration in your IDE (like visual studio, clion, or simply running `ctest`) because [CMake seems to copy the bin as if it was called from the directory where `add_test(...)` was called from](https://cmake.org/cmake/help/latest/prop_test/WORKING_DIRECTORY.html).

Since we don't use `add_test()`, at least not directly, then we'll have to go by the documentation for `gtest_discover_tests()` instead.

https://github.com/diasurgical/devilutionX/blob/bc4f43490f36cbc7478524bf8cb37e0afbd8c833/test/CMakeLists.txt#L51-L52

That said, it seems the `WORKING_DIRECTORY` property of that call works the same.

https://cmake.org/cmake/help/latest/module/GoogleTest.html#command:gtest_discover_tests
> Specifies the directory in which to run the discovered test cases. If this option is not provided, the current binary directory is used.

---

> I do wonder: should we not use the value of the current working directory as the base path, instead of the path to the binary? Either way, I haven't fixed it yet, and the solution seems to just run the test through the test explorer and let `ctest` do the magic.

We do use `CMAKE_CURRENT_BINARY_DIR` by default as the base path to deploy the test fixtures.

https://github.com/diasurgical/devilutionX/blob/bc4f43490f36cbc7478524bf8cb37e0afbd8c833/test/Fixtures.cmake#L99

https://github.com/diasurgical/devilutionX/blob/bc4f43490f36cbc7478524bf8cb37e0afbd8c833/test/Fixtures.cmake#L1-L3

That path should match the one `gtest_discover_tests()` is using as its `WORKING_DIRECTORY`. Since the executable is deployed to `CMAKE_RUNTIME_OUTPUT_DIRECTORY`, I do agree that it would make more sense if we used a relative path in the test exectuable instead.

However, I'm not really that familiar with ctest. It looks to me like CMake is generating a bunch of files in the format `*_test[1]_tests.cmake` that lists the `WORKING_DIRECTORY` property. This is presumably the working directory that CMake uses to execute the test. I'm not sure what Visual Studio, CLion, or ctest would use. Or does the bootstrapper within the test executable itself set up its own working directory before executing the test? 🤷 

---

Anyway, if you want to give it a shot, here's where we set pref path up to use relative paths in `pfile_write_hero`.

https://github.com/diasurgical/devilutionX/blob/bc4f43490f36cbc7478524bf8cb37e0afbd8c833/test/writehero_test.cpp#L372

matheusgomes28 commented:
@StephenCWills 

> Since we don't use add_test(), at least not directly, then we'll have to go by the documentation for gtest_discover_tests() instead.

If you look at the source code for `gtest_discover_tests()`, seems to be pretty much a wrapper around `add_test()` with some extra google test specific stuff, hence the same behaviour.

I kind of gave up trying to fix the prefix paths outside the ctest integration and just hardcoded the correct `pathPrefix(...)` in the test so I can debug. I'm pretty sure `ctest` changes the binary path itself by either copying the binary to the `WORKING_DIR` or some other way, then the SDL calls get the binary path dir, not the current working dir as the base directory.

Either way, I think I fixed the tests... turns out I removed one too many if checks on item locations.

AJenbo commented:
Looks good over all, this is a truly massive function so hard to tackle. It's fine if you do it in steps and just extract some of the obvious bits first.

matheusgomes28 commented:
Thanks @AJenbo . I'll removed some of the comments I just realised I left, then you are free to merge this.

kphoenix137 commented on Source/inv.cpp:
```diff
@@ -347,50 +491,89 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+
+std::optional<int8_t> GetPrevItemId(int slot, const Player &player, const Size &itemSize)
+{
+	int8_t item_something = 0;
```
Perhaps a better variable name? This is quite ambigious

matheusgomes28 replied:
I agree, hence why this feedback request! I guess it gets the item ID from the inv grid? I was not too sure on this one.

AJenbo replied:
Looks like `prevItemId` would have been a good name since that reflects the getter function name and this is the returned value.

kphoenix137 commented on Source/inv.cpp:
```diff
@@ -347,50 +491,89 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+		return;
+	}
+
+	auto const maybe_it = (il == ILOC_UNEQUIPABLE) ? GetPrevItemId(slot, player, itemSize) : 0;
```
Also a quite ambiguous variable name

matheusgomes28 replied:
I agree, hence why this feedback request! I guess it gets the item ID from the inv grid? I was not too sure on this one.

AJenbo commented:
Nice this is a good enough clean up that the function is now manageable to read :+1: 

I'll go over it in detail soon.

AJenbo commented on Source/inv.cpp:
```diff
@@ -405,155 +587,35 @@ void CheckInvPaste(Player &player, Point cursorPosition)
-		}
+	case ILOC_UNEQUIPABLE: {
+		ChangeInvItem(player, *maybe_item_id, slot, itemSize);
 		break;
+	}
```
would be nice to use the same break style now that the switch isn't so bloated.

AJenbo commented on Source/inv.cpp:
```diff
@@ -325,14 +325,158 @@ int FindTargetSlotUnderItemCursor(Point cursorPosition, Size itemSize)
+void ChangeBodyEquipment(int slot, Player &player, item_equip_type il)
 {
-	Size itemSize = GetInventorySize(player.HoldItem);
+	auto iLocToInvLoc = [&slot](item_equip_type loc) {
+		// Table?
+		switch (loc) {
+		case ILOC_HELM:
+			return INVLOC_HEAD;
+		case ILOC_RING:
+			return (slot == SLOTXY_RING_LEFT ? INVLOC_RING_LEFT : INVLOC_RING_RIGHT);
+		case ILOC_AMULET:
+			return INVLOC_AMULET;
+		case ILOC_ARMOR:
+			return INVLOC_CHEST;
+		default:
+			app_fatal("Unexpected equipment type");
+		}
+	};
+	const auto body_loc = iLocToInvLoc(il);
```
Since you do not need the lambda it self but just the result I would suggest executing it right after creating it:

```suggestion
	const auto body_loc = [&slot](item_equip_type loc) {
		// Table?
		switch (loc) {
		case ILOC_HELM:
			return INVLOC_HEAD;
		case ILOC_RING:
			return (slot == SLOTXY_RING_LEFT ? INVLOC_RING_LEFT : INVLOC_RING_RIGHT);
		case ILOC_AMULET:
			return INVLOC_AMULET;
		case ILOC_ARMOR:
			return INVLOC_CHEST;
		default:
			app_fatal("Unexpected equipment type");
		}
	}(il);
```

AJenbo commented on Source/inv.cpp:
```diff
@@ -347,50 +491,88 @@ void CheckInvPaste(Player &player, Point cursorPosition)
 		}
-	} else if (il == ILOC_BELT) {
+	} else {
+		auto const maybe_overlapped_item = CheckOverlappingItems(slot, player, itemSize, item_something);
+		if (!maybe_overlapped_item) {
+			return std::nullopt;
+		}
+		item_something = *maybe_overlapped_item;
```
We prefer having the type rather then auto when the type can't bee seen on the same line. This also removes the need for calling it "maybe", additionally the prefered style is `studdedCase` (local) or `CamelCase` (global) rather then `snake_case`, I know this isn't clear since a lot of code is mixed casing atm.
```suggestion
		const std::optional<int8_t> overlappingItemId = CheckOverlappingItems(slot, player, itemSize, item_something);
		if (!overlappingItemId) {
			return std::nullopt;
		}
		item_something = *overlappingItemId;
```

AJenbo commented on Source/inv.cpp:
```diff
@@ -347,50 +489,88 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+		return;
+	}
+
+	auto const maybe_item_id = (il == ILOC_UNEQUIPABLE) ? GetPrevItemId(slot, player, itemSize) : 0;
+	if ((il == ILOC_UNEQUIPABLE) && !maybe_item_id) {
 		return;
 	}
```
I'm not sure we are really winning anything by moving this up here. But maybe we could instead move it inside ChangeInvItem() since that is the only place this gets used. It would cause CalcPlrInv() to sometimes get called, but I don't think that is really an issue and if it is we can just have a return value for ChangeInvItem() that we check to see if we should exit the function early.
```suggestion
	std::optional<int8_t> maybe_item_id = 0;
	if (il == ILOC_UNEQUIPABLE) {
		maybe_item_id = GetPrevItemId(slot, player, itemSize);
		if (!maybe_item_id) return;
	}
```

AJenbo commented on Source/inv.cpp:
```diff
@@ -405,155 +586,32 @@ void CheckInvPaste(Player &player, Point cursorPosition)
-			player.HoldItem = previouslyEquippedItem;
-		}
+	case ILOC_ARMOR:
+		ChangeBodyEquipment(slot, player, il);
```
It would be nice if these new functions for changing equipment took there arguments in the same order: Player, Slot, Other, rather then being a bit random about it.

AJenbo commented on Source/inv.cpp:
```diff
@@ -347,53 +489,92 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+	return it;
+}
+
+std::optional<int8_t> GetPrevItemId(int slot, const Player &player, const Size &itemSize)
```
I don't really have anything against std::optional, but only positive values are valid returns from this so we can use -1 to indicate that nothing was found which allows a bit more direct use with out the need to check before passing the value on to the next user.

Some thing more relevant for this function and refactoring is that if you flip the if statements you can get rid of the else cases and instead make use of early returns. This takes a lot fewer lines and avoids the pyramid structure of the nested conditions.

This also reveals something odd in that CheckOverlappingItems() is always feed a 0 as the 4th argument as this value is never set by anything before it's called.

```cpp
int8_t GetPrevItemId(int slot, const Player &player, const Size &itemSize)
{
	if (player.HoldItem._itype != ItemType::Gold)
		return CheckOverlappingItems(slot, player, itemSize, 0);
	int8_t item_cell_begin = player.InvGrid[slot - SLOTXY_INV_FIRST];
	if (item_cell_begin == 0)
		return 0;
	if (item_cell_begin <= 0)
		return -item_cell_begin;
	if (player.InvList[item_cell_begin - 1]._itype != ItemType::Gold)
		return item_cell_begin;
}
```

AJenbo reviewed (changes requested):
When running this PR I'm unable to place items in the belt using the mouse. Could you look in to what is causing this?

AJenbo commented on Source/inv.cpp:
```diff
@@ -347,50 +537,28 @@ void CheckInvPaste(Player &player, Point cursorPosition)
 		if (!CanBePlacedOnBelt(player, player.HoldItem))
 			return;
-	} else if (desiredIl != il) {
+	}
+
+	if ((il != ILOC_UNEQUIPABLE) && (desired_loc != il)) {
```
Previously there was and else in between these, with out it it is now not possible to place items on the belt as `desired_loc` is never `ILOC_BELT`.
```suggestion
	} else if ((il != ILOC_UNEQUIPABLE) && (desired_loc != il)) {
```

AJenbo reviewed (approved):
I applied a few changes and did a bit further cleanups, hope you are ok with me doing so. Overall a good clean up, thanks for putting time an effort in to it.
